### PR TITLE
ASoC: sof_es8336: reduce pop noise on speaker

### PR DESCRIPTION
**Pull request by yangxiaohua2009**
The Speaker need to be turned on slightly behind the codec turned on.
It also need to be turned off slightly before the codec turned down.
Current code uses delay in DAPM_EVENT to do it but the mdelay delays the
DAPM itself and thus has no effect. A delayed_work is added to turn on the
speaker.
The speaker is turned off in .trigger since trigger is called slightly before
the DAPM events.

Signed-off-by: Zhu Ning <zhuning@everest-semi.com>

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

**Comment by yangxiaohua2009** [comment]
@plbossart Can you review it? Thanks.
Also, can two gpios be used by default?

File: sound/soc/intel/boards/sof_es8336.c [hunk 1]
```diff
@@ -122,19 +160,7 @@ static int sof_es8316_speaker_power_event(struct snd_soc_dapm_widget *w,
...
-
-	if (!(quirk & SOF_ES8336_HEADPHONE_GPIO))
-		return 0;
-
-	if (SND_SOC_DAPM_EVENT_ON(event))
-		msleep(70);
-
-	gpiod_set_value_cansleep(priv->gpio_headphone, priv->speaker_en);
-
+	queue_delayed_work(system_wq, &priv->pcm_pop_work, msecs_to_jiffies(70));
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c:165]: when you introduce any queuing, you need to make sure all the work is handled in the .remove function as well.

File: sound/soc/intel/boards/sof_es8336.c [hunk 2]
```diff
@@ -670,6 +716,20 @@ static int sof_es8336_probe(struct platform_device *pdev)
...
+		out_obj->integer.value ==  SOF8316_MIC_TYPE_DMIC_LOW )
+		quirk |= SOF_ES8336_ENABLE_DMIC;
+	else if (out_obj->integer.value ==  SOF8316_MIC_TYPE_AMIC_LIN2)
+		quirk |= SOC_ES8336_HEADSET_MIC1;
+	out_obj = acpi_evaluate_dsm(ACPI_HANDLE(codec_dev), &essx_guid, 1, SOF8316_HPMIC_TYPE, NULL);
+	dev_dbg(dev, "SOF8316_HPMIC_TYPE %llx", out_obj->integer.value);
+	out_obj = acpi_evaluate_dsm(ACPI_HANDLE(codec_dev), &essx_guid, 1, SOF8316_HPDET_INV, NULL);
+	dev_dbg(dev, "SOF8316_HPDET_INV %llx", out_obj->integer.value);
+	if (out_obj->integer.value ==  SOF8316_HPDET_INV_INVERTED)
+		quirk |= SOF_ES8336_JD_INVERTED;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c]: all of this seems to be a different functionality that needs to be in a different patch, not sure what this has to do with speaker pop noise?

File: sound/soc/intel/boards/sof_es8336.c [hunk 3]
```diff
@@ -111,6 +124,45 @@ static void log_quirks(struct device *dev)
...
+	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+	case SNDRV_PCM_TRIGGER_RESUME:
+		break;
+
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_STOP:
+		if(priv->speaker_en == false)
+			if(substream->stream == 0)
+				gpiod_set_value_cansleep(priv->gpio_speakers, true);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_es8336.c]: no idea what you are trying to do here.

**Review by plbossart — CHANGES_REQUESTED** [review]

File: sound/soc/intel/boards/sof_es8336.c (hunk 3 above)
Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c]: The speaker need to be turned off before dapm turned off, so they are turned off here. Is there any better way to do this?

Comment (plbossart) [reply sound/soc/intel/boards/sof_es8336.c]: can you clarify what level enables the speakers? Here you set the GPIO line to high, in other places you set it to speaker_en. Not clear at all

Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c]: The speaker is turned on by set it's value to false `gpiod_set_value_cansleep(priv->gpio_speakers, false);`
Because an 'active_low'  gpio is requested, and an 'active_low'  gpio turned off means high level, which turns on the speaker.

Comment (plbossart) [reply sound/soc/intel/boards/sof_es8336.c]: ok, it's really confusing that you are using speaker_en in a inverted way, this is far from self-explanatory.

**Comment by sokhapkin** [comment]
Any update on the issue? I'm experiencing this "pop noise" issue and would be glad to test a fix.

**Comment by yangxiaohua2009** [comment]
@sokhapkin Please apply the patch and recompile your kernel.

**Comment by sokhapkin** [comment]
> @sokhapkin Please apply the patch and recompile your kernel.

I did it already, the pop sound is still there.

File: sound/soc/intel/boards/sof_es8336.c (hunk 1 above)
Comment (yangxiaohua2009) [reply sound/soc/intel/boards/sof_es8336.c:165]: fixed

**Review by plbossart — COMMENTED** [review]
You probably want to send this to the alsa-devel mailing list @yangxiaohua2009, it's not clear to me if this is the right way to enforce such power-related sequences.

**Comment by plbossart** [comment]
sent upstream
